plan/predicate_push_down.go

```diff
@@ -45,10 +48,142 @@ func columnSubstitute(expr expression.Expression, schema expression.Schema, newE
 	return expr
 }
 
+// propagateConstant propagate constant values in a condition.
+// e.g. for condition: "a = b and b = c and c = a and a = 1";
+// we propagate constant as the following step:
+// first: "1 = b and b = c and c = 1 and a = 1";
+// next:  "1 = b and 1 = c and c = 1 and a = 1";
+// next:  "1 = b and 1 = c and 1 = 1 and a = 1";
+// next:  "1 = b and 1 = c and a = 1";
+
+// e.g for condition: "a = b and b = c and b = 2 and a = 1";
+// we propagate constant as the following step:
+// first: "a = 2 and 2 = c and b = 2 and a = 1";
+// next:  "a = 2 and 2 = c and b = 2 and 2 = 1";
+// next:  "0"
+func propagateConstant(conditions []expression.Expression) []expression.Expression {
+	if len(conditions) == 0 {
+		return conditions
+	}
+	isHandled := make([]bool, len(conditions))
+	equalities := make(map[*expression.Column]*expression.Constant, 0)
```

Comment:
Use *Column as map key is not robust. Same column may not have same physical address. Judge it by FromID and PositionID.